common/src/main/java/com/github/khanshoaib3/minecraft_access/features/read_crosshair/CrosshairNarrator.java

```diff
@@ -0,0 +1,31 @@
+package com.github.khanshoaib3.minecraft_access.features.read_crosshair;
+
+import net.minecraft.client.world.ClientWorld;
+import net.minecraft.util.hit.HitResult;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface CrosshairNarrator {
+    @NotNull HitResult rayCast();
+
+    /**
+     * Determines weather the player is looking at the same block or entity.
+     * {@link #narrate(ClientWorld, boolean)} will only be called
+     * if this method returns a different non-null value than its previous call.
+     * A value of null indicates that the player is not currently looking at any block
+     * and thus {@link #narrate(ClientWorld, boolean)} should not be called.
+     * @param world The current world the player is in
+     * @param speakSide Weather the block side the player is pointing at should be narrated
+     * @return Any object comparable with {@link java.util.Objects#equals(Object, Object)} or null
+     */
+    @Nullable Object deduplication(@NotNull ClientWorld world, boolean speakSide, boolean speakConsecutiveBlocks);
+
+    /**
+     * Generates a description of the block or entity the player is currently looking at to be read out.
+     * This will never be called if {@link #deduplication(ClientWorld, boolean, boolean)} returns null.
+     * @param world The current world the player is in
+     * @param speakSide Weather the block side the player is pointing at should be narrated
+     * @return A non-null string to be read to the player
+     */
+    @NotNull String narrate(@NotNull ClientWorld world, boolean speakSide);
```

Comment:
`narrate`'s param can be the return value of `deduplication`, then we don't need `rayCast`, and this method can be extract to `NarratorUtils`, I guess?

Reply:
`rayCast` exists because I noticed that Jade's ray casting was subtly different and I wanted everything to line up with Jade if it's active. I don't think passing `deduplication`'s return value to `narrate` would be very useful and it would be a bit of a mess with generics because `deduplicatiom`'s return type is currently `Object`. All methods in the `CrosshairNarrator` interface are completely different between both implementations so i don't think it makes sense to move anything to `NarratorUtils`.